### PR TITLE
fix Ghost#9656; Fix having duplicate streams for logging

### PR DESCRIPTION
**Pull request by abdallahalsamman**
Closes https://github.com/TryGhost/Ghost/issues/9656

### CODE REVIEW

**Comment by kirrg001** [comment]
Thanks 🙏🏽

I wonder why


```    
"logging": {
        "level": "info",
        "rotation": {
            "enabled": true,
            "count": 1,
            "period": "100000ms"
        },
        "transports": ["file", "stdout"]
    }
```

has not shown duplicated entries? Any idea?

**Comment by abdallahalsamman** [comment]
Good question! I tried debugging it and couldn't come to a conclusion

What do we do in this case?

**Comment by kirrg001** [comment]
So weird. I'll have a look at this weirdo on Monday. Thanks for your help - really nice :)

**Comment by kirrg001** [comment]
So i think i know why this is behaved a little weird.

If you just define

```
logging: {
  rotation: true
}
```

It falls back to

> { enabled: true, period: '1d', count: 10 }

That means both streams (file + rotation) operate on the same file for 1 day. This results in duplicated entries.

But if you define

```
"logging": {
        "rotation": {
            "enabled": true,
            "count": 1,
            "period": "100000ms"
        }
}
```

Then the file stream always writes into the oldest log file, but the rotation stream into the latest log file. So if you watch the oldest log file, you will see entries are getting added when rotation is enabled, which should not happen. That is probably because: bunyan creates a write stream to `yourblog.log`, but this file get's rotated to e.g. `log.1` and the file write stream writes into `log.1`, then `log.2` etc.

So i think we can merge this :) THanks!

**Review by kirrg001 — APPROVED** [review]

**Comment by kirrg001** [comment]
I just had a thought - maybe there is a different fix for that, because it's weird that the rotation stream writes log entries.

**Comment by kirrg001** [comment]
Yeah. The problem is that our logger iterates over all available streams (contains file stream, rotation stream) and calls `logger.log`, which will produce a log entry.

https://github.com/TryGhost/Ignition/blob/master/lib/logging/GhostLogger.js#L304

Here we iterate. A condition is needed to exclude rotation streams :)

**Comment by abdallahalsamman** [comment]
@kirrg001 Nice work finding out where the bug was!

I pushed again, the fix you suggested works like a charm

Edit: I've added a condition to check for rotation streams and skip them while iterating on streams

File: lib/logging/GhostLogger.js [hunk 1]
```diff
@@ -306,6 +306,8 @@ class GhostLogger {
             // because it would result in duplicate logs
             if (type === 'error' && logger.name === 'stdout' && includes(this.transports, 'stderr')) {
                 return;
+            } else if ( logger.name.indexOf('rotation') > -1 ) {
+                return;
```
Comment (kirrg001) [inline lib/logging/GhostLogger.js]: Thanks for the update. Could you please add a description?
Or add a link to this PR discussion. Thanks!

Comment (abdallahalsamman) [reply lib/logging/GhostLogger.js]: Sure, updated comment https://github.com/TryGhost/Ignition/pull/59#issuecomment-394352539

Comment (kirrg001) [reply lib/logging/GhostLogger.js]: Sorry i meant in the code base - a little description why we exclude file rotation.

**Comment by abdallahalsamman** [comment]
@kirrg001 I've added a link to this PR in the code, are we good to merge?

**Comment by kirrg001** [comment]
@evexoio Thanks 🤠

**Pull request merged**